cpp/src/arrow/type.cc

```diff
@@ -686,6 +690,67 @@ std::shared_ptr<Schema> schema(std::vector<std::shared_ptr<Field>>&& fields,
   return std::make_shared<Schema>(std::move(fields), metadata);
 }
 
+Status UnifySchemas(const std::vector<std::shared_ptr<Schema>>& schemas,
+                    std::shared_ptr<Schema>* out) {
+  if (schemas.empty()) {
+    return Status::Invalid("Must provide at least one schema to unify.");
+  }
+
+  std::vector<std::shared_ptr<Field>> fields = schemas[0]->fields();
+  std::unordered_map<std::string, size_t> field_name_to_index;
+  for (size_t i = 0; i < fields.size(); ++i) {
+    if (!field_name_to_index.emplace(fields[i]->name(), i).second) {
+      return Status::Invalid(
+          "UnifySchemas does not support duplicate field names in the schema: ",
+          fields[i]->name());
+    }
+  }
+
+  auto update_field = [](const std::shared_ptr<Field>& other,
```

Comment:
I think this lambda is complex enough to be extracted as a static helper function. Eventually (follow up PR, maybe) it'd be good to have individual unit tests for it as well

Reply:
Made a static function.